### PR TITLE
K8SPS-245: generate missing passwords in user secret

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPS-245](https://jira.percona.com/browse/K8SPS-245) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPS-245

**DESCRIPTION**
---
**Problem:**
*When user deletes one of the passwords in secrets, cluster fails*

**Cause:**
*We are unable to set an empty password for user*

**Solution:**
*Operator should generate missing passwords and update user secret*

**CHECKLIST**
---
**Jira**
- [x] Is the Jira ticket created and referenced properly?
- [x] Does the Jira ticket have the proper statuses for documentation (`Needs Doc`) and QA (`Needs QA`)?
- [x] Does the Jira ticket link to the proper milestone (Fix Version field)?

**Tests**
- [x] Is an E2E test/test case added for the new feature/change?
- [x] Are unit tests added where appropriate?

**Config/Logging/Testability**
- [x] Are all needed new/changed options added to default YAML files?
- [x] Are the manifests (crd/bundle) regenerated if needed?
- [x] Did we add proper logging messages for operator actions?
- [x] Did we ensure compatibility with the previous version or cluster upgrade process?
- [x] Does the change support oldest and newest supported PS version?
- [x] Does the change support oldest and newest supported Kubernetes version?

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/auto-config.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/gr-init-deploy.log) |
| gr-scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/gr-scaling.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/sidecars.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/config.log) |
| gr-demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/gr-demand-backup.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/limits.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/version-service.log) |
| config-router | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/config-router.log) |
| tls-cert-manager | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/tls-cert-manager.log) |
| gr-tls-cert-manager | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/gr-tls-cert-manager.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/scaling.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/one-pod.log) |
| haproxy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/haproxy.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/users.log) |
| async-ignore-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/async-ignore-annotations.log) |
| gr-ignore-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-311/5cbca7b/gr-ignore-annotations.log) |
| We run 22 out of 22|

commit: https://github.com/percona/percona-server-mysql-operator/pull/311/commits/5cbca7be40b92ee2af2fe1032e8919c65aca2981
image: `perconalab/percona-server-mysql-operator:PR-311-5cbca7b`

**Review by egegunes — DISMISSED** [review]

File: pkg/controller/ps/user_test.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	apiv1alpha1 "github.com/percona/percona-server-mysql-operator/api/v1alpha1"
+	"github.com/percona/percona-server-mysql-operator/pkg/secret"
+)
+
+func TestEnsureUserSecrets(t *testing.T) {
```
Comment (inelpandzic) [inline pkg/controller/ps/user_test.go:18]: Could you try to write those with Gingko?

**Review by hors — DISMISSED** [review]

File: pkg/controller/ps/user_test.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	ctx := context.Background()
+	secretsName := "some-secret"
+	ns := "some-namespace"
+	tests := []struct {
+		name   string
+		cr     *apiv1alpha1.PerconaServerMySQL
+		secret *corev1.Secret
+	}{
+		{
+			name: "Test without user secret",
```
[the remark this answers is not in the document]
Comment (pooknull) [reply pkg/controller/ps/user_test.go]: https://github.com/percona/percona-server-mysql-operator/pull/311/commits/f0c3c4d87b2e21ae232477fbc2441667c92dfa54

File: pkg/controller/ps/user_test.go (hunk 1 above)
Comment (pooknull) [reply pkg/controller/ps/user_test.go:18]: I will do it in another PR

**Review by inelpandzic — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Pull request merged**